Bluetooth: Mesh: Access model callback struct

trond-snekvik opened the pull request:
Adds a structure of callbacks for each model instance. This allows for more flexible model implementations, that can interact with the Mesh stack without going through the application.

The motivation for this change is to improve the user experience when adding modular, reusable model implementations. The intention is to make it as easy to mix and match models as it is with drivers, where the user only has to list their models in the composition data to enable them. 

Added callbacks for init and model user data persistent storage. The init callback replaces the internal lookup for foundation models. Also adds variation of the `BT_MESH_MODEL` macro with a callback parameter, without breaking the backward compatibility of the old macro.

Looking for some input, particularly in these areas:
Any other callbacks we should add right away? 
Are there better ways to enable model instance settings storage? 
Should the model persistent storage be scheduled like the other mesh settings?

trond-snekvik committed:
Bluetooth: Mesh: Add model callback structure

Adds a structure of callbacks for each model instance. This allows for
more flexible model implementations, that can interact with the Mesh
stack without going through the application.

For now, only an init callback is added, replacing the init mechanism in
the foundation models. The init callback does not provide the primary
flag that used to be in the foundation model callbacks, but replaces
this with an inline function in access.h.

Signed-off-by: Trond Einar Snekvik <Trond.Einar.Snekvik@nordicsemi.no>

trond-snekvik committed:
Bluetooth: Mesh: Add persistent model user data

Adds user data file in the model settings, allowing each model instance
to store some application specific data without having to reinvent the
settings path encoding for model element/ID combinations. Exposes the
settings_handler interface in the model callback structure and adds a
data store function.

Signed-off-by: Trond Einar Snekvik <Trond.Einar.Snekvik@nordicsemi.no>

jhedberg commented:
@trond-snekvik looks good to me, but there seems to have crept in some conflict with `access.h` - you'll need to rebase against latest master and fix it before I can merge this.

jhedberg commented:
One thing that I started wondering about: what about node reset? How are you planning to handle that? Should the respective model data be deleted from flash in that case, or do you just leave it hanging around there until the node gets provisioned again? Should there be some new callback for this. We do have a global reset callback currently though.

trond-snekvik commented:
That's a good point. I think we should delete the model data on reset, and as a consequence, each model should probably be notified of this change. I can add this to this PR.

trond-snekvik committed:
Bluetooth: Mesh: Add model reset callback

Adds additional model callback that gets called on node_reset. Will also
erase any user data when this happens.

Signed-off-by: Trond Einar Snekvik <Trond.Einar.Snekvik@nordicsemi.no>

trond-snekvik commented on subsys/bluetooth/mesh/cfg_srv.c:
```diff
@@ -3273,21 +3278,33 @@ int bt_mesh_cfg_srv_init(struct bt_mesh_model *model, bool primary)
+			bt_mesh_store_mod_sub(mod);
+		}
+
+		bt_mesh_model_data_store(mod, vnd, NULL, 0);
```
If I'm reading the settings module correctly, this'll actually generate 0-length entries for models that don't have any user data. I couldn't find a good way to get around this though, do you have any suggestions, or do you think it's fine @jhedberg?

jhedberg replied:
A new flag to `model->flags` to indicate if data was ever stored to or retrieved from flash? Similar to what I've tried to do here for pairing keys: https://github.com/jhedberg/zephyr/commit/7e22ca25bc53a97331876bff97c232dcc20a0333

trond-snekvik replied:
Of course 🤦‍♂ Made the change.
